connection.go

```diff
@@ -72,11 +72,17 @@ Loop:
 				timer.Reset(i.timeout)
 			}
 		case <-expired:
-			for _, stream := range i.conn.streams {
-				stream.Reset()
-			}
+			i.conn.streamCond.L.Lock()
+			streams := i.conn.streams
+			i.conn.streams = make(map[spdy.StreamId]*Stream)
+			i.conn.streamCond.Broadcast()
+			i.conn.streamCond.L.Unlock()
+			go func() {
+				for _, stream := range streams {
+					stream.resetStream()
+				}
+			}()
 			i.conn.Close()
```

Comment:
Inside the goroutine after the resets, if that works for you?
On Fri, Feb 27, 2015 at 7:12 PM Derek McGowan <notifications@github.com>
wrote:

> In connection.go
> <https://github.com/docker/spdystream/pull/47#discussion_r25549366>:
>
> > @@ -72,11 +72,17 @@ Loop:
> >  				timer.Reset(i.timeout)
> >  			}
> >  		case <-expired:
> > -			for _, stream := range i.conn.streams {
> > -				stream.Reset()
> > -			}
> > +			i.conn.streamCond.L.Lock()
> > +			streams := i.conn.streams
> > +			i.conn.streams = make(map[spdy.StreamId]*Stream)
> > +			i.conn.streamCond.Broadcast()
> > +			i.conn.streamCond.L.Unlock()
> > +			go func() {
> > +				for _, stream := range streams {
> > +					stream.resetStream()
> > +				}
> > +			}()
> >  			i.conn.Close()
>
> Suggesting i.conn.Close is inside the goroutine before the resets? I don't
> like that hanging is a possibility on Close but I see it is only because
> the resetChan cannot be read from while this call is done. I will make that
> update, my main concern was i.conn.streams was updated before Close.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/spdystream/pull/47/files#r25549366>.
>